backend/src/telegram/utils/callback/handlers/contact.ts

```diff
@@ -16,46 +22,56 @@ const logger = loggerWithLabel(module)
 const upsertTelegramSubscriber = async (
   phoneNumber: string,
   telegramId: number
-): Promise<boolean> => {
+): Promise<UpsertTelegramSubscriberResult> => {
   const logMeta = {
     phoneNumber,
     telegramId,
     action: 'upsertTelegramSubscriber',
   }
+
   // Some Telegram clients send pre-prefixed phone numbers
   if (!phoneNumber.startsWith('+')) {
     phoneNumber = `+${phoneNumber}`
   }
 
-  /**
-   * Insert a telegram id and phone number, if that telegram id doesn't exist.
-   * Otherwise, if the new phone number does not exist,
-   * update the existing phone number associated with that telegram id with the new phone number,
-   * to maintain a 1-1 mapping.
-   */
-  const result = await TelegramSubscriber?.sequelize?.query(
-    `
-        INSERT INTO telegram_subscribers (phone_number, telegram_id, created_at, updated_at)
-        VALUES (:phoneNumber, :telegramId, clock_timestamp(), clock_timestamp())
-        ON CONFLICT (telegram_id) DO UPDATE
-        SET phone_number = :phoneNumber, updated_at = clock_timestamp()
-        WHERE NOT telegram_subscribers.phone_number = :phoneNumber
-      `,
-    {
-      replacements: {
-        phoneNumber,
-        telegramId,
-      },
+  const result = await TelegramSubscriber.sequelize?.transaction(
+    async (transaction) => {
+      /**
+       * Split into 2 scenarios:
+       * 1. a row with matching phone_number or telegram_id: update the row
+       * 2. no row matches: insert a new row
+       *
+       * Note: we have to use a raw query to update the phone_number.
+       * It is a primary key, and sequelize doesn't allow us to update
+       * primary keys in sequelize models.
+       */
```

Comment:
I tried a few fixes which used sequelize model methods (`findOrCreate`, `findOne` then `save`). However, they didn't work for updating `phone_number` due to this issue. I kept them as separate commits in this branch just for reference.

In the end, the simplest and probably the cleanest way to fix this seems to be the current iteration: trying to update, else insert.